src/libsystemd/sd-device/device-trigger.c

```diff
@@ -0,0 +1,483 @@
+
+                        r = set_put_strdup(trigger->settle_set, settle_key);
+                        if (r < 0)
+                                return r;
+                }
+        }
+
+        return ret;
+}
+
+int sd_device_trigger_set_pre_trigger_callback(sd_device_trigger *trigger, sd_device_trigger_callback_t cb, void *userdata) {
+        trigger->pre_trigger_cb = cb;
+        trigger->pre_trigger_cb_userdata = userdata;
+
+        return 0;
+}
+
+int sd_device_trigger_set_settle_callback(sd_device_trigger *trigger, sd_device_trigger_callback_t cb, void *userdata) {
+        trigger->settle_cb = cb;
+        trigger->settle_cb_userdata = userdata;
+
+        return 0;
+}
+
+_public_ int sd_device_trigger_execute(sd_device_trigger *trigger) {
+        int r;
+
+        assert_return(trigger, -EINVAL);
+
+        if (trigger->executing)
+                return -EPERM;
+
+        r = device_trigger_ensure_uuid_set(trigger);
+        if (r < 0)
+                return r;
+
+        r = device_trigger_execute_common(trigger);
+
+        device_trigger_discard_uuid(trigger);
+
+        return r;
+}
+
+static int device_trigger_monitor_handler(sd_device_monitor *monitor, sd_device *device, void *userdata) {
+        sd_device_trigger *trigger = userdata;
+        const char *syspath;
+        const char *uuid;
+        _cleanup_free_ char *settle_key = NULL;
+        char *found_key;
+        int r;
+
+        r = sd_device_get_syspath(device, &syspath);
+        if (r < 0)
+                return r;
+
+        r = sd_device_get_property_value(device, SYNTH_UUID_KEY, &uuid);
+        if (r < 0) {
+               if (r == -ENOENT)
+                       /* ignore uevents which are not synthetic */
+                       return 0;
+               return r;
+        }
+
+        if (!streq(uuid, trigger->uuid))
+                /* we are interested in synthetic uevents with specific UUID only */
+                return 0;
+
+        settle_key = strjoin(trigger->uuid, syspath);
+        if (!settle_key)
+                return -ENOMEM;
+
+        found_key = set_remove(trigger->settle_set, settle_key);
+        if (!found_key)
+                return 0;
+
+        free(found_key);
+
+        log_debug("sd-device-trigger: %s: %s: Settled", trigger->uuid, syspath);
+
+         if (trigger->settle_cb)
+                 (void) trigger->settle_cb(trigger, device, trigger->settle_cb_userdata);
+
+        if (set_isempty(trigger->settle_set))
+                return sd_event_exit(sd_device_monitor_get_event(monitor), 0);
+
+        return 0;
+}
+
+static int device_trigger_timeout_handler(sd_event_source *s, uint64_t usec, void *userdata) {
+        sd_device_trigger *trigger = userdata;
+
+        log_debug("sd-device-trigger: %s: Timed out", trigger->uuid);
+
+        return sd_event_exit(sd_event_source_get_event(s), -ETIMEDOUT);
+}
+
+_public_ int sd_device_trigger_execute_with_settle(sd_device_trigger *trigger, unsigned int timeout_sec) {
```

Comment:
uh, a synchronous, blocking call in sd-device? that doesn't look good and would be a first. if people want blocking behaviour, fine, but that should happen in the apps using the lib, no here.